trie/fog.py

```diff
@@ -0,0 +1,284 @@
+import ast
+from itertools import zip_longest
+from typing import (
+    Any,
+    Sequence,
+    Tuple,
+)
+
+from eth_utils import (
+    to_tuple,
+    ValidationError,
+)
+from sortedcontainers import SortedSet
+
+from trie.typing import (
+    Nibbles,
+    NibblesInput,
+    RawHexaryNode,
+)
+from trie.utils.nodes import (
+    key_starts_with,
+)
+
+
+class HexaryTrieFog:
+    """
+    Keeps track of which parts of a trie have been verified to exist.
+
+    Named after "fog of war" popular in video games like... Red Alert? IDK, I'm old.
+
+    Object is immutable. Any changes, like marking a key prefix as complete, will
+    return a new HexaryTrieFog object.
+    """
+    _unexplored_prefixes: SortedSet
+
+    # INVARIANT: No unexplored prefix may start with another unexplored prefix
+    #   For example, _unexplored_prefixes may not be {(1, 2), (1, 2, 3)}.
+
+    def __init__(self) -> None:
+        # Always start without knowing anything about a trie. The only unexplored
+        #   prefix is the root prefix: (), which means the whole trie is unexplored.
+        self._unexplored_prefixes = SortedSet({()})
+
+    def __repr__(self) -> str:
+        return f"HexaryTrieFog<{self._unexplored_prefixes!r}>"
+
+    @property
+    def is_complete(self) -> bool:
+        return len(self._unexplored_prefixes) == 0
+
+    def explore(
+            self,
+            old_prefix: NibblesInput,
+            sub_segments_input: Sequence[NibblesInput]) -> 'HexaryTrieFog':
+        """
+        The fog lifts from the old prefix. This call returns a HexaryTrieFog that narrows
+        down the unexplored key prefixes. from the old prefix to the indicated children.
+
+        For example, if only the key prefix 0x12 is unexplored, then calling
+        explore((1, 2), ((3,), (0xe, 0xf))) would mark large swaths of 0x12 explored, leaving only
+        two prefixes as unknown: 0x123 and 0x12ef. To continue exploring those prefixes, navigate
+        to them using traverse() or traverse_from().
+
+        The sub_segments_input may be empty, which means the old prefix has been fully explored.
+        """
+        sub_segments = [Nibbles(segment) for segment in sub_segments_input]
+        new_fog_prefixes = self._unexplored_prefixes.copy()
+
+        try:
+            new_fog_prefixes.remove(old_prefix)
+        except KeyError:
+            raise ValidationError(f"Old parent {old_prefix} not found in {new_fog_prefixes!r}")
+
+        if len(set(sub_segments)) != len(sub_segments):
```

Comment:
If `sub_segments_input` was a `Set` type then we wouldn't have to perform this validation check.

Reply:
For now I'm :-1: on the change, because I like that the argument can be passed in directly from a `HexaryTrieNode.sub_segments`. Making `sub_segments_input` a set would mean doing one of:
- casting `set(node.sub_segments)` at any call site
- changing `HexaryTrieNode.sub_segments` to a set, but it's a data structure that is really nice to keep immutable